### PR TITLE
Add RLE compressed binary image support

### PR DESCRIPTION
**Pull request by XuNeo**
### Description of the feature or fix

Add flag to image header to indicate the data is compressed.
Separate definition of `data`, using union to include compressed image data struct.

Added test for RLE compressed image.

Update python script to support RLE compressed image.
Usage:
```
  --compress {NONE,RLE}
                        Binary data compress method, default to NONE
```

### Checkpoints
- [ ] Run `code-format.py` from the scripts folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [ ] Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed
- [ ] Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant. 
- [ ] Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- [ ] If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/release/v8.3/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/release/v8.3/Kconfig).
 
Be sure the following conventions are followed:
- [ ] Follow the [Styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Prefer `enum`s instead of macros. If inevitable to use `define`s export them with `LV_EXPORT_CONST_INT(defined_value)` right after the `define`.
- [ ] In function arguments prefer `type name[]` declaration for array parameters instead of `type * name`
- [ ] Use typed pointers instead of `void *` pointers
- [ ] Do not `malloc` into a static or global variables. Instead declare the variable in `lv_global_t` structure in [`lv_global.h`](https://github.com/lvgl/lvgl/blob/master/src/core/lv_global.h) and mark the variable with `(LV_GLOBAL_DEFAULT()->variable)` when it's used. See a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#memory-management).
- [ ] Widget constructor must follow the `lv_<widget_name>_create(lv_obj_t * parent)` pattern.
- [ ] Widget members function must start with `lv_<module_name>` and should receive `lv_obj_t *` as first argument which is a pointer to widget object itself.  
- [ ] `struct`s should be used via an API and not modified directly via their elements.
- [ ] `struct` APIs should follow the widgets' conventions. That is to receive a pointer to the `struct` as the first argument, and the prefix of the `struct` name should be used as the prefix of the function name too (e.g.  `lv_disp_set_default(lv_disp_t * disp)`)
- [ ] Functions and `struct`s which are not part of the public API must begin with underscore in order to mark them as "private".
- [ ] Arguments must be named in H files too.
- [ ] To register and use callbacks one of the following needs to be followed (see a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#callbacks)):
  - For both the registration function and the callback pass a pointer to a `struct` as the first argument. The `struct` must contain `void * user_data` field.
  - The last argument of the registration function must be `void * user_data` and the same `user_data` needs to be passed as the last argument of the callback.
  - Callback types not following these conventions should end with `xcb_t`.

### CODE REVIEW

**Comment by XuNeo** [comment]
Hi Gabor,
The CI reports error but I cannot reproduce locally. Could you help me on how to reproduce it?

**Comment by kisvegabor** [comment]
It fails for me locally too. This is how image_format_rle_compressed_err.png looks like:  
image_format_rle_compressed_err

It seems get_info is not working as the size not set.

File: src/draw/lv_image_buf.h [hunk 1]
```diff
@@ -113,9 +126,20 @@ typedef struct {
...
-    const uint8_t * data;   /**< Pointer to the data of the image*/
+    union {
+        const uint8_t * data;   /**< Pointer to the data of the image*/
+        struct _lv_image_compressed_t {
+            uint32_t method: 4; /*Compression method, see `lv_image_compress_t`*/
+            uint32_t reserved : 28;  /*Reserved to be used later*/
+            uint32_t compressed_size;  /*Compressed data size in byte*/
+            uint32_t decompressed_size;  /*Decompressed data size in byte*/
+            const uint8_t * data; /*Compressed data*/
+        } compressed;
```
Comment (kisvegabor) [inline src/draw/lv_image_buf.h]: See my comment [here](https://github.com/lvgl/lvgl/issues/4869#issuecomment-1829425792).

Comment (XuNeo) [reply src/draw/lv_image_buf.h]: > In an other issue we were talking about having an lv_image_dsc_t for the compressed/original image, and an other lv_draw_buf_t for the decoded image. Can we assume that the decoder can set the uncompressed data size during decoding? If it's required for the decoder I would rather store in the compressed image's header. I feel like it's a custom data required only for custom formats, so it should not be exposed as public.

> @XuNeo One more note: nameless unions are C11 standard we are on C99. We can consider jumping to C11, but it requires serious consideration and research.

You are correct. Let's move the compressed data format inside lv_bin_decoder.c and use C99.

Comment (XuNeo) [reply src/draw/lv_image_buf.h:137]: > In an other issue we were talking about having an lv_image_dsc_t for the compressed/original image, and an other lv_draw_buf_t for the decoded image. Can we assume that the decoder can set the uncompressed data size during decoding? If it's required for the decoder I would rather store in the compressed image's header. I feel like it's a custom data required only for custom formats, so it should not be exposed as public.

> @XuNeo One more note: nameless unions are C11 standard we are on C99. We can consider jumping to C11, but it requires serious consideration and research.

You are correct. Let's move the compressed data format inside lv_bin_decoder.c and use C99.

**Comment by XuNeo** [comment]
Hi @kisvegabor,
This PR is ready for review. Could you check it? Thank you.

**Review by kisvegabor — COMMENTED** [review]
Looks good, thank you!

**Pull request merged**